arm64-assembly-optimization.md

```diff
@@ -0,0 +1,120 @@
+
+## Instruction selection and scheduling
+
+ARM provides several Software Optimization Guides (SWOG) for each CPU:
+- [Graviton2 - Neoverse-N1 SWOG](https://developer.arm.com/documentation/swog309707/a/)
+- [Graviton3 - Neoverse-V1 SWOG](https://developer.arm.com/documentation/PJDOC-466751330-9685/0101/)
+
+The SWOG documents for each instruction the latency (number of cycles it takes
+to finish the execution of an instruction) and the throughput (the number of
+similar instructions that can execute in parallel.) The SWOG also documents the
+execution units that can execute an instruction. The information provided by the
+SWOG is sometimes encoded in compilers (GCC and LLVM) under the form of specific
+tuning flags for different CPUs. Tuning flags allow compilers to produce a good
+instruction scheduling and a good instruction selection.
+
+LLVM has a tool [llvm-mca](https://www.llvm.org/docs/CommandGuide/llvm-mca.html)
+that allows software developers to see on their input assembly code how the
+compiler reasons about the execution of instructions by a specific processor.
+
+## Splitting data dependence chains
+
+To increase instruction level parallelism it is possible to duplicate
+computations. The redundant computations can execute in parallel as they belong
+to different dependence chains.
+
+For example, it is possible to issue several loads to break data dependent
+instructions.  The data becomes available earlier and avoids stalls due to data
+dependences.  Instead of using a load and an extract instructions:
+
+```
+ld1  {v0.16b}, [x1]
+// v0 is available after ld1 finishes execution
+ext  v1.16b, v0.16b, v0.16b, #1   // x1 + 1
+// v1 is available only after ld1 and ext finish execution
+```
+we can issue two independent loads that can execute in parallel:
+```
+ld1  {v0.16b}, [x1]
+ld1  {v1.16b}, [x1, #1]     // x1 + 1
+// v0 and v1 are available after the two independent ld1 instructions finish execution
+// Both Graviton2 and Graviton3 will execute the two loads in parallel.
+```
+
+## Modulo scheduling
+
+Data loaded for iteration `i` can be used in the next iteration `i + 1` by
+keeping it in a register.  The registers used for modulo scheduling are pre-load
+before the loop starts, and each iteration saves the data needed in the next
+iteration in those registers.
+
+## Optimizing a dot product by a constant vector
+
+When the data is known ahead of execution time, it is possible to specialize the
+code based on the data. For example in the x265 encoder a filter is using the
+constant coefficients that need to be multiplied by data read from memory and
+then summing up all the results. The C code does the following:
+
+```
+const int16_t g_lumaFilter[4][NTAPS_LUMA] =
+{
+    {  0, 0,   0, 64,  0,   0, 0,  0 },
+    { -1, 4, -10, 58, 17,  -5, 1,  0 },
+    { -1, 4, -11, 40, 40, -11, 4, -1 },
+    {  0, 1,  -5, 17, 58, -10, 4, -1 }
+};
+[...]
+    const int16_t* coeff = g_lumaFilter[coeffIdx];
```

Comment:
It would probably be good to to rename this constant array for the example code.